File: ai/no_easy_prisoner.rb
# Should recognize when our move is foolish...

require_relative "heuristic"
require_relative "hunter"

class NoEasyPrisoner < Heuristic

  def initialize(player)
    super
    set_as_negative
    @enemy_hunter = Hunter.new(player,true)
  end

  def init_color
    super
    @enemy_hunter.init_color
  end

  def eval_move(i,j)
    # NB: snapback is handled in hunter; here we just notice the sacrifice of a stone, which will
    # be balanced by the profit measured by hunter (e.g. lose 1 but kill 3).
    stone = Stone.play_at(@goban,i,j,@color)
    g = stone.group
    score = 0
    if g.lives == 1
      score = - g.stones.size
      $log.debug("NoEasyPrisoner says #{i},#{j} is plain foolish (#{score})") if $debug
    elsif g.lives == 2
      $log.debug("NoEasyPrisoner asking Hunter to look at #{i},#{j}") if $debug
      if @enemy_hunter.escaping_atari_is_caught?(stone)
        score = - g.stones.size
        $log.debug("NoEasyPrisoner (backed by Hunter) says #{i},#{j} is foolish  (#{score})") if $debug
      end
    end
    Stone.undo(@goban)
    return score
  end

end

File: js/ai/NoEasyPrisoner.js
//Translated from no_easy_prisoner.rb using babyruby2js
'use strict';

var main = require('../main');

var Grid = require('../Grid');
var Heuristic = require('./Heuristic');
var inherits = require('util').inherits;
var Stone = require('../Stone');

var sOK = main.sOK;


/** @class Should recognize when our move is foolish... */
function NoEasyPrisoner(player) {
    Heuristic.call(this, player);
    this.executioner = player.getHeuristic('Executioner');
    this.hunter = null;
}
inherits(NoEasyPrisoner, Heuristic);
module.exports = NoEasyPrisoner;


NoEasyPrisoner.prototype.evalBoard = function (stateYx, scoreYx) {
    if (!this.hunter) this.hunter = this.player.getHeuristic('Hunter');
    var myScoreYx = this.scoreGrid.yx;
    var executionerYx = this.executioner.scoreGrid.yx;
    for (var j = 1; j <= this.gsize; j++) {
        for (var i = 1; i <= this.gsize; i++) {
            if (stateYx[j][i] < sOK) continue;
            var score = myScoreYx[j][i] = this.evalMove(i, j);
            if (score === 0) continue;
            // Remove score given by Executioner if this move does not work
            // var execScore = executionerYx[j][i];
            // if (execScore < -score) score -= execScore;
            scoreYx[j][i] += score;
        }
    }
};

NoEasyPrisoner.prototype.evalMove = function (i, j) {
    // NB: snapback is handled in hunter; here we just notice the sacrifice of a stone, which will
    // be balanced by the profit measured by hunter (e.g. lose 1 but kill 3).
    var stone = Stone.playAt(this.goban, i, j, this.color);
    var g = stone.group;
    var score = 0, move;
    if (main.debug) move = Grid.xy2move(i, j);
    if (g.lives === 1) {
        score = - this.groupThreat(g);
        if (main.debug) main.log.debug('NoEasyPrisoner says ' + move + ' is plain foolish (' + score + ')');
    } else if (g.lives === 2) {
        if (main.debug) main.log.debug('NoEasyPrisoner asking Hunter to look at ' + move);
        if (this.hunter.escapingAtariIsCaught(stone)) {
            score = - this.groupThreat(g, true);
            if (main.debug) main.log.debug('NoEasyPrisoner (backed by Hunter) says ' + move + ' is foolish  (' + score + ')');
        }
    }
    Stone.undo(this.goban);
    return score;
};
